docs/wrapper-layer.md

```diff
@@ -0,0 +1,53 @@
+
+A *logical API*, on the other hand, is one that is closely tied to specific to functional units, like currencies or measurement units.  A logical API function should require a specific hunk of data and perform a specific task.
+
+Although ergonomic APIs correspond with what users of i18n intend to do, from an implementation standpoint, they are often quite heavy-weight, pulling in many dependencies that may not be required for all call sites.  Intl.NumberFormat is an example: it supports currency formatting, which requires a lot of code and data, but when formatting a number that's not a currency, you shouldn't need to carry all that extra code and data.
+
+The goal of the *wrapper layer* is to provide an ergonomic API for clients, but map it to the logical API under the hood.
+
+![Ergonomic to Logical API](assets/ergonomic-logical.svg)
+
+For example, in the above diagram, a client uses the ergonomic Intl.NumberFormat API with options to enable compact notation in long form.  The wrapper layer performs a bit of logic to figure out that this combination of options requires two functional units: plural rules and compact decimal format.  It then delegates to those two functions in the logical API.
+
+## Key Requirements for the Wrapper Layer
+
+### Host Language
+
+Since the ergonomic API depends on the host language (e.g., Objective-C calling into OmnICU via an FFI), the wrapper layer should be written in each individual host language, mapping to the logical API exported by the OmnICU library.
+
+This allows different languages to choose different styles for passing options into the library; for example, a Builder pattern could be used in Java, and an options bag could be used in JavaScript.
+
+### Code Slicing
+
+The wrapper layer should be suitable for tools to perform automatic dead code elimination.  By resolving options at compile time, the set of logical APIs required for a certain application can be known in time to perform static analysis; unused logical APIs can then be removed, reducing code and data size.
+
+In languages that support function inlining, the wrapper layer should generally be inlined into the call site, such that dead code elimination can tailor the set of logical APIs to each individual call site.
+
+Rust has great support for dead code elimination.  In languages where tooling for dead code elimination is not as strong, this should be taken into consideration when designing the ergonomic API specific to that language.  The ergonomic API might need to have a more logical flavor in order for code slicing to work as intended.
+
+### Data Loading
+
+In accordance with [principles.md](principles.md), the core library should not perform I/O; instead, data should be passed via a dynamic DataProvider ([data-pipeline.md](data-pipeline.md)).
+
+How does this fit into the wrapper layer model?
+
+1. The Ergonomic API receives a Data Provider.
```

Comment:
1. This is not obvious from your diagram
2. Asking end client to always provide data provider in the API looks clunky bcs. it's mostly unrelated to functionality at hand @hagbard 

I'll open new issue to scope this better.

Reply:
Added issue #30